packages/sdk-react/src/core/queries/useReceivables.ts

```diff
@@ -357,19 +362,42 @@ export const usePDFReceivableById = (
   }
 ) => {
   const { monite } = useMoniteContext();
+  const queryClient = useQueryClient();
 
   return useQuery<ReceivableFileUrl, ApiError>({
     queryKey: receivablesQueryKeys.pdf(receivableId),
+    queryFn: async () => {
+      const response = await monite.api.receivable.getPdfLink(receivableId);
+
+      if (!response.file_url) {
+        /**
+         * We have to flush current query data to show the error
+         *  but not previous data
+         */
+        queryClient.setQueryData(receivablesQueryKeys.pdf(receivableId), null);
```

Comment:
Can refactor the custom result handling from `queryFn`? I don't see the need to use such a complicated pattern, it seems it would be easier to use `refetchInterval`

Reply:
Can't say that `refetchInterval` is nicer but I tried but the problem is that `refetchInterval` will call BE again and again even if BE fails many times.

So in my current approach (at least for me) everything is straightforward. Query function should describe what should be done to fetch the data. So in our case if there is no `file_url` we have to fetch the data again but no more than 4 times just to avoid any BE overload.

So in `retry` we could tackle it. If we constantly have 4 errors in a row we will throw an error inside `QueryFn` which is `File URL is not provided` and this error will be provided to public API inside `error`.

But can't implement so for the `refetchInterval` because there is no way to provide an error to the customer.
[Also I see that `retry` is designed for this logic](https://tanstack.com/query/latest/docs/framework/react/guides/query-retries#retry-delay) to describe should we `retry` or not based on condition. And if we throw an error in `queryFn` we could manage it

Reply:
Your approach complicates future Qraft integration. Of course, a custom `queryFn` could be specified for its hooks as well. But why, when you can do the same with the approach I suggested?

Reply:
Okay, but how would you fix the problem of stopping iterating after N times and providing an error to the customer? As I have in the current implementation? To stop pulling BE repeatedly if they provide no PDF and show an error to the client at the same time.

Reply:
Don't you think that's a pretty rare case? Maybe, on the contrary, it's worth trying time after time to get the link if it's empty? Empty link !== API error.

Reply:
> Don't you think that's a pretty rare case?

I think it's important to cover all possible scenarios we never know when PDF will be generated. Maybe it will never will be generated because the customer provided some wrong data and our BE may never return the data. So answering your question: I think it's a rare case but it doesn't mean that we shouldn't handle it.

> it's worth trying time after time to get the link if it's empty

Yep, in the current approach, we try 4 times and if it still fails we show an error and provide a button (see in `.gif` file) when the user may manually click and re-fetch it.

---

So I see a few problems with `refetchInterval`:
1. If BE responds with an error still call BE over and over again no matter what
2. If BE responds with an empty `file_url` we also call BE over and over again no matter what

As I can see by [Tanstack documentation](https://tanstack.com/query/latest/docs/framework/react/guides/query-retries#retry-delay) `retry` is designed specifically for `queryFn` errors because it provides `failureCount` and `error` and based on that we could decide should we `retry` or not. But `refetch` is a general terminology responsible for refetching data not retrying failed ones.

But if it's so important for you to avoid `retry` and write it inside `refetchInterval` would you mind helping me with these 2 problems that I mentioned above? How do we handle errors and avoid any BE requests?

Reply:
It seems we just have different perspectives on how requests from APIs should be handled.